internal/engine/cache.go

```diff
@@ -0,0 +1,189 @@
+	gvk, err := apiutil.GVKForObject(obj, c.scheme)
+	if err != nil {
+		return errors.Wrap(err, "cannot determine group, version, and kind of supplied object")
+	}
+
+	c.mx.RLock()
+	if _, active := c.active[gvk]; active {
+		defer c.mx.RUnlock()
+		return c.Cache.Get(ctx, key, obj, opts...)
+	}
+	c.mx.RUnlock()
+
+	c.mx.Lock()
+	defer c.mx.Unlock()
+	c.active[gvk] = true
+	return c.Cache.Get(ctx, key, obj, opts...)
+}
+
+// List retrieves list of objects for a given namespace and list options. On a
+// successful call, Items field in the list will be populated with the result
+// returned from the server.
+//
+// Listing objects marks the informer for the object's GVK active.
+func (c *InformerTrackingCache) List(ctx context.Context, list client.ObjectList, opts ...client.ListOption) error {
+	gvk, err := apiutil.GVKForObject(list, c.scheme)
+	if err != nil {
+		return errors.Wrap(err, "cannot determine group, version, and kind of supplied object")
+	}
+	gvk.Kind = strings.TrimSuffix(gvk.Kind, "List")
+
+	c.mx.RLock()
+	if _, active := c.active[gvk]; active {
+		defer c.mx.RUnlock()
+		return c.Cache.List(ctx, list, opts...)
+	}
+	c.mx.RUnlock()
+
+	c.mx.Lock()
+	defer c.mx.Unlock()
+	c.active[gvk] = true
+	return c.Cache.List(ctx, list, opts...)
+}
+
+// GetInformer fetches or constructs an informer for the given object that
+// corresponds to a single API kind and resource.
+//
+// Getting an informer for an object marks the informer as active.
+func (c *InformerTrackingCache) GetInformer(ctx context.Context, obj client.Object, opts ...cache.InformerGetOption) (cache.Informer, error) {
+	gvk, err := apiutil.GVKForObject(obj, c.scheme)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot determine group, version, and kind of supplied object")
+	}
+
+	c.mx.RLock()
+	if _, active := c.active[gvk]; active {
+		defer c.mx.RUnlock()
+		return c.Cache.GetInformer(ctx, obj, opts...)
+	}
+	c.mx.RUnlock()
+
+	c.mx.Lock()
+	defer c.mx.Unlock()
+	c.active[gvk] = true
+	return c.Cache.GetInformer(ctx, obj, opts...)
+}
+
+// GetInformerForKind is similar to GetInformer, except that it takes a
+// group-version-kind, instead of the underlying object.
+//
+// Getting an informer marks the informer as active.
+func (c *InformerTrackingCache) GetInformerForKind(ctx context.Context, gvk schema.GroupVersionKind, opts ...cache.InformerGetOption) (cache.Informer, error) {
+	c.mx.RLock()
+	if _, active := c.active[gvk]; active {
+		defer c.mx.RUnlock()
+		return c.Cache.GetInformerForKind(ctx, gvk, opts...)
+	}
+	c.mx.RUnlock()
+
+	c.mx.Lock()
+	defer c.mx.Unlock()
+	c.active[gvk] = true
+	return c.Cache.GetInformerForKind(ctx, gvk, opts...)
+}
+
+// RemoveInformer removes an informer entry and stops it if it was running.
+//
+// Removing an informer marks the informer as inactive.
+func (c *InformerTrackingCache) RemoveInformer(ctx context.Context, obj client.Object) error {
```

Comment:
thought: a smoke (unit) test might help to detect when the `cache.Cache` interface changes. The overridden methods are must be exhaustive for a given gvk. So there is risk that this breaks and will be hard to debug.

Reply:
Another approach that could work would be to stop embedding `cache.Cache`. Instead wrap one and implement all of its methods. That way we'll stop satisfying the interface if it changes. Will follow up in another PR.